hibernate-core/src/test/java/org/hibernate/orm/test/annotations/onetoone/OverrideOneToOneJoinColumnTest.java

```diff
@@ -0,0 +1,206 @@
+package org.hibernate.orm.test.annotations.onetoone;
+
+import org.hibernate.AnnotationException;
+import org.hibernate.boot.Metadata;
+import org.hibernate.boot.MetadataSources;
+import org.hibernate.boot.model.naming.Identifier;
+import org.hibernate.boot.registry.StandardServiceRegistry;
+import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
+import org.hibernate.mapping.ForeignKey;
+import org.hibernate.mapping.Table;
+
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.orm.junit.BaseUnitTest;
+
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.AssociationOverride;
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.Embedded;
+import jakarta.persistence.Entity;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.OneToOne;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/**
+ * @author Aresnii Skvortsov
+ */
+@TestForIssue(jiraKey = "HHH-4384")
+@BaseUnitTest
+public class OverrideOneToOneJoinColumnTest {
+
+	@Test
+	public void allowIfJoinColumnIsAbsent() {
+		StandardServiceRegistry ssr = new StandardServiceRegistryBuilder().build();
+		try {
+			Metadata metadata = new MetadataSources( ssr )
+					.addAnnotatedClass( Person.class )
+					.addAnnotatedClass( State.class )
+					.buildMetadata();
+
+			Table personTable = metadata.getDatabase().getDefaultNamespace().locateTable( Identifier.toIdentifier(
+					"PERSON_TABLE" ) );
+			ForeignKey foreignKey = personTable.getForeignKeyIterator().next();
+
+			assertEquals(
+					"PERSON_ADDRESS_STATE",
+					foreignKey.getColumn( 0 ).getName(),
+					"Overridden join column name should be applied"
+			);
+
+		}
+		finally {
+			StandardServiceRegistryBuilder.destroy( ssr );
+		}
+	}
+
+	@Test
+	public void disallowOnSideWithMappedBy() {
+		StandardServiceRegistry ssr = new StandardServiceRegistryBuilder().build();
+		try {
+			new MetadataSources( ssr )
+					.addAnnotatedClass( Employee.class )
+					.addAnnotatedClass( PartTimeEmployee.class )
+					.addAnnotatedClass( Desk.class )
+					.buildMetadata();
+			fail( "Should disallow @JoinColumn override on side with mappedBy" );
+		}
+		catch (AnnotationException ex) {
+			assertTrue(
+					"Should disallow exactly because of @JoinColumn override on side with mappedBy",
+					ex
+							.getMessage()
+							.startsWith( "Illegal attempt to define a @JoinColumn with a mappedBy association:" )
+			);
+		}
```

Comment:
The above boilerplate could be simplified by JUnit 5's `assertThrows()` method as following:
```
final AnnotationException ex = Assertions.assertThrows(AnnotationException.class, () -> new MetadataSources( ssr )
					.addAnnotatedClass( Employee.class )
					.addAnnotatedClass( PartTimeEmployee.class )
					.addAnnotatedClass( Desk.class )
					.buildMetadata());
assertTrue(
					"Should disallow exactly because of @JoinColumn override on side with mappedBy",
					ex.getMessage().startsWith( "Illegal attempt to define a @JoinColumn with a mappedBy association:" )
			);
```
The benefits include:
* more compact and elegant
* avoid the common error to forget the `fail()` statement (happy path would slip through)

One good example of lambda usage shines in JUnit 5.
Need not to be changed; just for communication purpose. In my current big bank contract, so many testing cases lack the 'fail()' statement ending up with failure to catch the happy paths.

Reply:
Thanks @NathanQingyangXu , applied your suggestion!